### PR TITLE
feat: add `--http.corsdomain`

### LINKED ISSUE #1230: Add CORS domain support
### Describe the feature

Since the HTTP server is accessible from any local application, additional protection has been built into the server to prevent misuse of the API from web pages.

To enable access to the API from a web page (for example, to use the online IDE Remix), the server must be configured to accept cross-origin requests.

## TODO
* add `http.corsdomain` CLI args ref #1221 
* Integrate CORS layer in rpc-builder ref https://github.com/paritytech/jsonrpsee/blob/f6f4a0845d7c94dfb9befe625b2a779b3a40a127/examples/examples/cors_server.rs

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Tirthnp**
Closes #1230

### CODE REVIEW

**Comment by Tirthnp** [comment]
I have created a draft version as it's a first pass. I just want to know if I am doing it right. I think we need to add support for `--http.corsdomain *`  and `--http.cordomain`.  I think we need to have an enum **RpcCorsConfig**. Can someone please give me some guidance on it?

**Comment by onbjerg** [comment]
Hey! When opening up a PR that closes an issue, you can use special words to link the PR and the issue, which auto-closes the issue when it is merged. For example:

Closes #1 or fixes #1 

It helps keep the issue tracker clean :smile: Thanks for taking this on!

File: crates/rpc/rpc-builder/src/lib.rs [hunk 1]
```diff
@@ -472,6 +477,31 @@ impl RpcServerConfig {
         self
     }
 
+    // Confiugures the http server with cors domain
+    pub fn with_http_cors_domain(mut self, config: ServerBuilder, domains: Vec<&str>) -> Self {
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:481]: I would rather have this as a separate field (cors_domans: Option<...>)and then this is created in `build`

File: bin/reth/src/node/mod.rs [hunk 2]
```diff
@@ -141,15 +141,28 @@ impl Command {
...
+            let domains_vec = domains.split(",").collect::<Vec<&str>>();
+            let _rpc_server = reth_rpc_builder::launch(
+                ShareableDatabase::new(db.clone()),
+                reth_transaction_pool::test_utils::testing_pool(),
+                network.clone(),
+                TransportRpcModuleConfig::default()
+                    .with_http(vec![RethRpcModule::Admin, RethRpcModule::Eth]),
+                RpcServerConfig::default().with_http_cors_domain(Default::default(),domains_vec),
+            )
+            .await?;
```
Comment (mattsse) [inline bin/reth/src/node/mod.rs:154]: we should leave this entirely up to the rpc builder.
It should handle `Option<domain>`

**Review by mattsse — CHANGES_REQUESTED** [review]
great!

left some suggestions

**Comment by Tirthnp** [comment]
I have made some changes as per the comments but I am not exactly sure how to test them

File: bin/reth/src/node/mod.rs [hunk 3]
```diff
@@ -155,7 +155,8 @@ impl Command {
             network.clone(),
             TransportRpcModuleConfig::default()
                 .with_http(vec![RethRpcModule::Admin, RethRpcModule::Eth]),
-            RpcServerConfig::default().with_http(Default::default()),
+            RpcServerConfig::default()
+                .with_http(Default::default(), self.rpc.http_corsdomain.as_ref()),
```
Comment (mattsse) [inline bin/reth/src/node/mod.rs:159]: the corsdomain should be configured in a separate function.
with_http should only take the server config imo.

File: crates/rpc/rpc-builder/src/lib.rs [hunk 4]
```diff
@@ -500,6 +503,8 @@ where
 pub struct RpcServerConfig {
     /// Configs for JSON-RPC Http.
     http_server_config: Option<ServerBuilder>,
+    /// Configs for JSON-RPC that allow Corsdomains
+    http_cors_domain_server_config: Option<ServerBuilder<Stack<CorsLayer, Identity>>>,
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:507]: separate for cors domain only

File: crates/rpc/rpc-builder/src/lib.rs [hunk 5]
```diff
@@ -529,10 +534,41 @@ impl RpcServerConfig {
...
+                "*" => {
+                    let cors = CorsLayer::new()
+                        .allow_methods([Method::GET, Method::POST])
+                        .allow_origin(Any)
+                        .allow_headers(Any);
+
+                    let middleware = tower::ServiceBuilder::new().layer(cors);
+                    self.http_cors_domain_server_config =
+                        Some(config.set_middleware(middleware).http_only());
+                }
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:550]: all of this needs to be set when the server is started

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks!

left some suggestions

File: crates/rpc/rpc-builder/src/lib.rs (hunk 1 above)
Comment (Tirthnp) [reply crates/rpc/rpc-builder/src/lib.rs]: Could you please elaborate on this? I am a bit confused

File: crates/rpc/rpc-builder/src/lib.rs (hunk 5 above)
Comment (gakonst) [reply crates/rpc/rpc-builder/src/lib.rs:550]: yeah @Tirthnp what you want here is to just set the config variable, and then in the build/start function you actually consume them

File: crates/rpc/rpc-builder/src/lib.rs [hunk 6]
```diff
@@ -529,10 +534,41 @@ impl RpcServerConfig {
     pub fn ipc(config: IpcServerBuilder) -> Self {
         Self::default().with_ipc(config)
     }
-
     /// Configures the http server
-    pub fn with_http(mut self, config: ServerBuilder) -> Self {
-        self.http_server_config = Some(config.http_only());
+    pub fn with_http(mut self, config: ServerBuilder, domains: Option<&String>) -> Self {
```
Comment (gakonst) [inline crates/rpc/rpc-builder/src/lib.rs:538]: The `domains` method should be separate, so one should be able to do `with_http(config).with_cors(Some("*"))`

**Review by gakonst — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc-builder/src/lib.rs [hunk 7]
```diff
@@ -609,9 +618,45 @@ impl RpcServerConfig {
...
+
+                        let origins = domains_vec
+                            .into_iter()
+                            .map(|domain| domain.parse().unwrap())
+                            .collect::<Vec<HeaderValue>>();
+
+                        cors = Some(CorsLayer::new()
+                            .allow_methods([Method::GET, Method::POST])
+                            .allow_origin(origins)
+                            .allow_headers(Any));
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:642]: please extract this to a standalone function that returns `Result<Option<CorsLayer>`

File: crates/rpc/rpc-builder/src/lib.rs [hunk 8]
```diff
@@ -609,9 +618,45 @@ impl RpcServerConfig {
...
+                            .allow_origin(Any)
+                            .allow_headers(Any));
+                    } 
+                    "" => {}
+                    _ => {
+                        let domains_vec = domains.split(",").collect::<Vec<&str>>();
+
+                        let origins = domains_vec
+                            .into_iter()
+                            .map(|domain| domain.parse().unwrap())
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:636]: can be combined into `domains.split(",").into_iter()`

File: crates/rpc/rpc-builder/src/lib.rs [hunk 9]
```diff
@@ -609,9 +618,45 @@ impl RpcServerConfig {
...
+                            .collect::<Vec<HeaderValue>>();
+
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:638]: once  move to function, collect into Result:
```suggestion
                            .collect::<Result<Vec<HeaderValue>,_>();

```

File: crates/rpc/rpc-builder/src/lib.rs [hunk 10]
```diff
@@ -609,9 +618,45 @@ impl RpcServerConfig {
...
+                        cors = Some(CorsLayer::new()
+                            .allow_methods([Method::GET, Method::POST])
+                            .allow_origin(origins)
+                            .allow_headers(Any));
+                    }
+                }   
+            }
+
+            match cors {
+                Some(cors) => {
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:648]: nit:
use `if let Some() else` instead of match

File: crates/rpc/rpc-builder/src/lib.rs [hunk 11]
```diff
@@ -742,12 +787,19 @@ pub struct RpcServer {
...
     /// ipc server
     ipc: Option<IpcServer>,
 }
+/// Http Servers Enum
+pub enum HttpServer {
+    /// Http server
+    Plain(Server),
+    /// Http server with cors
+    WithCors(Server<Stack<CorsLayer, Identity>>),
+}
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:802]: this is okay for now, perhaps we find a better solution but for now this is ok

**Review by mattsse — CHANGES_REQUESTED** [review]
great progress.

mostly style nits, otherwise this is pretty close.

File: crates/rpc/rpc-builder/src/lib.rs (hunk 11 above)
Comment (gakonst) [reply crates/rpc/rpc-builder/src/lib.rs:807]: We could dyn dispatch

File: crates/rpc/rpc-builder/src/lib.rs [hunk 12]
```diff
@@ -635,6 +653,38 @@ impl RpcServerConfig {
...
+                            .allow_origin(Any)
+                            .allow_headers(Any),
+                    );
+                }
+                "" => {}
+                _ => {
+                    let origins = domains
+                        .split(",")
+                        .map(|domain| domain.parse::<HeaderValue>())
+                        .collect::<Result<Vec<HeaderValue>, _>>();
```
Comment (gakonst) [inline crates/rpc/rpc-builder/src/lib.rs:674]: Nit I'd add a CoraDomains type that impls FromStr and includes this logic, and do Option<CorsDomains> in the upstream cli code

File: crates/rpc/rpc-builder/src/lib.rs [hunk 13]
```diff
@@ -609,9 +618,18 @@ impl RpcServerConfig {
...
+                let middleware = tower::ServiceBuilder::new().layer(cors);
+                let http_server =
+                    builder.set_middleware(middleware).build(http_socket_addr).await?;
+                server.http_local_addr = http_server.local_addr().ok();
+                server.http = Some(HttpServer::WithCors(http_server));
+            } else {
+                let http_server = builder.build(http_socket_addr).await?;
+                server.http_local_addr = http_server.local_addr().ok();
+                server.http = Some(HttpServer::Plain(http_server));
+            }
```
Comment (gakonst) [inline crates/rpc/rpc-builder/src/lib.rs:632]: Towers middlewares are so powerful. We should investigate what else is available @mattsse may let us do nice rate limiting etc features in the node

**Review by gakonst — APPROVED** [review]
Good with me, couple nits which we can do in follow up

**Review by mattsse — APPROVED** [review]
lgtm,

please enable edits by maintainer next PR

**Comment by Tirthnp** [comment]
> please enable edits by maintainer next PR

I will, Thanks you so much for all the help :)

**Pull request merged**